mirage/unikernel.ml

```diff
@@ -428,10 +454,43 @@ module Main (N : Mirage_net.S) (ASSETS : Mirage_kv.RO) = struct
               (Blocklist.blocked_domains (Resolver.primary_data resolver))
           in
           Some (`Content (Dashboard.dashboard_layout ~content (), None))
-      | `POST, "/blocklist/add" ->
-          (* TODO: we can't get the body here oh no *)
-          None
-      | `POST, s when String.starts_with s ~prefix:"/blocklist/delete/" -> (
+      | `POST (content_type_header, data), "/blocklist/add" -> (
+          let content_type =
+            Option.fold ~none:"application/x-www-form-urlencoded\r\n"
+              ~some:(fun s -> s ^ "\r\n")
+              content_type_header
+          in
+          match Multipart_form.Content_type.of_string content_type with
+          | Error (`Msg e) ->
+              Logs.err (fun m -> m "Bad content-type header: %s" e);
+              Some (`Bad_request ("/blocklist", None))
+          | Ok ct -> (
+              match Multipart_form.of_string_to_list data ct with
+              | Error (`Msg e) ->
+                  Logs.err (fun m -> m "Error parsing form: %s" e);
+                  Some (`Bad_request ("/blocklist", None))
+              | Ok (m, assoc) -> (
+                  let multipart_body, _r = to_map ~assoc m in
+                  match Map.find_opt "domain" multipart_body with
+                  | None ->
+                      Logs.err (fun m -> m "No domain field in form");
+                      Some (`Bad_request ("/blocklist", None))
+                  | Some (_, domain_str) -> (
+                      match Domain_name.of_string domain_str with
+                      | Error (`Msg e) ->
+                          Logs.debug (fun m ->
```

Comment:
why on debug level? Feels like we should think about the log levels a bit more...